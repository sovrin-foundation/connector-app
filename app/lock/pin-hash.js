// @flow
import PBKDF2 from 'react-native-pbkdf2'
import { NativeModules, Platform } from 'react-native'
import { captureError } from '../services/error/error-handler'

const { RNRandomBytes } = NativeModules

const generateKey = (password: string, salt: string) =>
  PBKDF2.derivationKey(password, salt, 1000)

export const generateSalt = async () => {
  const numBytes = 32
  return new Promise(function(resolve, reject) {
    RNRandomBytes.randomBytes(numBytes, (err: any, bytes: string) => {
      if (err) {
        reject(err)
      } else {
        if (Platform.OS === 'android') {
          resolve(bytes.slice(0, -1))
        } else {
          resolve(bytes)
        }
      }
    })
  })
}

export async function pinHash(pin: string, salt: string) {
  try {
    const key = await generateKey(pin, salt)
    if (__DEV__) {
      console.log('pinHash: salt: ', salt)
      console.log('pinHash: key: ', key)
    }
    //TODO: This is hack due to android
    return key.substring(0, 16)
  } catch (e) {
    console.error(`pinHash: ${e}`)
    captureError(new Error(`pinHash: ${e}`))
    return null
  }
}
