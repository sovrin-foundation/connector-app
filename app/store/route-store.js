// @flow
import { splashScreenRoute } from '../common/route-constants'
import type { RouteStore } from './type-store'
import type { InitialTestAction } from '../common/type-common'

export type RouteStoreAction = typeof handleRouteUpdate | InitialTestAction

const initialState = {
  currentScreen: splashScreenRoute,
  timeStamp: new Date().getTime(),
}

export const ROUTE_UPDATE = 'ROUTE_UPDATE'

export const handleRouteUpdate = (currentScreen: string) => ({
  type: ROUTE_UPDATE,
  currentScreen,
})

export default function routeReducer(
  state: RouteStore = initialState,
  action: RouteStoreAction
) {
  switch (action.type) {
    case ROUTE_UPDATE:
      return {
        currentScreen: action.currentScreen,
        timeStamp: new Date().getTime(),
      }
    default:
      return state
  }
}
