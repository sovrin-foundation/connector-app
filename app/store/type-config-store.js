// @flow
import type { CustomError } from '../common/type-common'
import {
  splashScreenRoute,
  lockSelectionRoute,
  switchEnvironmentRoute,
  eulaRoute,
  restoreRoute,
} from '../common/route-constants'

export const SERVER_ENVIRONMENT = {
  DEMO: 'DEMO',
  SANDBOX: 'SANDBOX',
  STAGING: 'STAGING',
  DEVELOPMENT: 'DEVELOPMENT',
  QATEST1: 'QATEST1',
  QATEST2: 'QATEST2',
  DEVRC: 'DEVRC',
  QARC: 'QARC',
}

export const __uniqueId = '__uniqueId'

export type ServerEnvironment = $Keys<typeof SERVER_ENVIRONMENT>

export const APP_INSTALLED = 'APP_INSTALLED'
export type AppInstalledSuccessAction = {
  type: typeof APP_INSTALLED,
}

export const HYDRATED = 'HYDRATED'
export type HydratedAction = {
  type: typeof HYDRATED,
}

export const INITIALIZED = 'INITIALIZED'
export type InitializedAction = {
  type: typeof INITIALIZED,
}
export const ALREADY_INSTALLED_RESULT = 'ALREADY_INSTALLED_RESULT'
export type AlreadyInstalledAction = {
  type: typeof ALREADY_INSTALLED_RESULT,
  isAlreadyInstalled: boolean,
}

export const SERVER_ENVIRONMENT_CHANGED = 'SERVER_ENVIRONMENT_CHANGED'
export type ServerEnvironmentChangedAction = {
  type: typeof SERVER_ENVIRONMENT_CHANGED,
  serverEnvironment: ServerEnvironment,
}

export const SWITCH_ENVIRONMENT = 'SWITCH_ENVIRONMENT'
export type ChangeEnvironment = {
  agencyUrl: string,
  poolConfig: string,
  agencyDID: string,
  agencyVerificationKey: string,
}

export type SwitchEnvironmentAction = {
  type: typeof SWITCH_ENVIRONMENT,
} & ChangeEnvironment

export const SWITCH_ERROR_ALERTS = 'SWITCH_ERROR_ALERTS'
export type SwitchErrorAlertsAction = {
  type: typeof SWITCH_ERROR_ALERTS,
}

export const TOGGLE_ERROR_ALERTS = 'TOGGLE_ERROR_ALERTS'
export type ToggleErrorAlertsAction = {
  type: typeof TOGGLE_ERROR_ALERTS,
  isShowErrorAlert: boolean,
}

export const SAVE_SWITCH_ENVIRONMENT_DETAIL_FAIL =
  'SAVE_SWITCH_ENVIRONMENT_DETAIL_FAIL'
export type SaveSwitchEnvironmentDetailFailAction = {
  type: typeof SAVE_SWITCH_ENVIRONMENT_DETAIL_FAIL,
  error: CustomError,
}

export const HYDRATE_SWITCH_ENVIRONMENT_DETAIL_FAIL =
  'HYDRATE_SWITCH_ENVIRONMENT_DETAIL_FAIL'

export type HydrateSwitchEnvironmentDetailFailAction = {
  type: typeof HYDRATE_SWITCH_ENVIRONMENT_DETAIL_FAIL,
  error: CustomError,
}

export const CHANGE_ENVIRONMENT_VIA_URL = 'CHANGE_ENVIRONMENT_VIA_URL'
export type ChangeEnvironmentUrlAction = {
  type: typeof CHANGE_ENVIRONMENT_VIA_URL,
  url: string,
}

export const VCX_INIT_NOT_STARTED = 'VCX_INIT_NOT_STARTED'

export const VCX_INIT_START = 'VCX_INIT_START'
export type VcxInitStartAction = {
  type: typeof VCX_INIT_START,
}

export const VCX_INIT_SUCCESS = 'VCX_INIT_SUCCESS'
export type VcxInitSuccessAction = {
  type: typeof VCX_INIT_SUCCESS,
}

export const VCX_INIT_FAIL = 'VCX_INIT_FAIL'
export type VcxInitFailAction = {
  type: typeof VCX_INIT_FAIL,
  error: CustomError,
}

export const USE_VCX = 'USE_VCX'
export type UseVcxAction = {
  type: typeof USE_VCX,
}

export type ConfigAction =
  | HydratedAction
  | InitializedAction
  | AppInstalledSuccessAction
  | AlreadyInstalledAction
  | ServerEnvironmentChangedAction
  | SwitchErrorAlertsAction
  | ToggleErrorAlertsAction
  | SwitchEnvironmentAction
  | SaveSwitchEnvironmentDetailFailAction
  | HydrateSwitchEnvironmentDetailFailAction
  | ChangeEnvironmentUrlAction
  | VcxInitStartAction
  | VcxInitFailAction
  | VcxInitSuccessAction
  | UseVcxAction

export type VcxInitializationState =
  | typeof VCX_INIT_NOT_STARTED
  | typeof VCX_INIT_START
  | typeof VCX_INIT_SUCCESS
  | typeof VCX_INIT_FAIL

export type AgencyPoolConfig = {
  agencyUrl: string,
  agencyDID: string,
  agencyVerificationKey: string,
  poolConfig: string,
}

export type ConfigStore = {
  isAlreadyInstalled: boolean,
  isHydrated: boolean,
  showErrorAlerts: boolean,
  vcxInitializationState: VcxInitializationState,
  vcxInitializationError: null | CustomError,
  isInitialized: boolean,
} & AgencyPoolConfig

export const STORAGE_KEY_SWITCHED_ENVIRONMENT_DETAIL =
  'STORAGE_KEY_SWITCHED_ENVIRONMENT_DETAIL'

export const ERROR_SAVE_SWITCH_ENVIRONMENT = {
  code: 'CS-000',
  message: 'Failed to store switched environment details: ',
}

export const ERROR_HYDRATE_SWITCH_ENVIRONMENT = {
  code: 'CS-001',
  message: 'Failed to hydrate switched environment details: ',
}

export const schemaDownloadedEnvironmentDetails = {
  type: 'object',
  properties: {
    agencyUrl: {
      type: 'string',
    },
    agencyDID: {
      type: 'string',
    },
    agencyVerificationKey: {
      type: 'string',
    },
    poolConfig: {
      type: 'string',
    },
  },
  required: ['agencyDID', 'agencyUrl', 'agencyVerificationKey', 'poolConfig'],
}

export const MESSAGE_FAIL_ENVIRONMENT_SWITCH_TITLE =
  'Failed to switch environment'

export const ERROR_NO_WALLET_NAME = 'Wallet name not found'

export const MESSAGE_FAIL_ENVIRONMENT_SWITCH_INVALID_DATA = (url: string) =>
  `Data returned by ${url} is not valid as per our requirements.`

export const MESSAGE_FAIL_ENVIRONMENT_SWITCH_ERROR = (message: string) =>
  `Failed to switch environment due to following error. ${message}`

export const MESSAGE_SUCCESS_ENVIRONMENT_SWITCH_TITLE = 'Success'

export const MESSAGE_SUCCESS_ENVIRONMENT_SWITCH_DESCRIPTION =
  'Environment switched successfully.'

export const ERROR_VCX_INIT_FAIL = (message: string) => ({
  code: `CS-002`,
  message: `Failed to init vcx: ${message}`,
})

export const ERROR_VCX_PROVISION_FAIL = (message: string) => ({
  code: `CS-003`,
  message: `Failed to provision vcx: ${message}`,
})

export const UNSAFE_SCREENS_TO_DOWNLOAD_SMS = [
  splashScreenRoute,
  lockSelectionRoute,
  switchEnvironmentRoute,
  eulaRoute,
  restoreRoute,
]

export type DownloadedMessage = {
  statusCode: string,
  payload?: ?any,
  senderDID: string,
  uid: string,
  type: string,
  refMsgId?: ?string,
  deliveryDetails: Array<any>,
  decryptedPayload?: string,
}
export type DownloadedConnectionMessages = {
  pairwiseDID: string,
  msgs: Array<DownloadedMessage>,
}
export type DownloadedConnectionsWithMessages = Array<
  DownloadedConnectionMessages
>
export type AcknowledgeServerData = Array<{
  pairwiseDID: string,
  uids: Array<string>,
}>

export const MESSAGE_RESPONSE_CODE = {
  MESSAGE_CREATED: 'MS-101',
  MESSAGE_SENT: 'MS-102',
  MESSAGE_PENDING: 'MS-103',
  MESSAGE_ACCEPTED: 'MS-104',
  MESSAGE_REJECTED: 'MS-105',
  MESSAGE_ANSWERED: 'MS-106',
}

export type MessageClaimOfferDetails = {|
  claim_id: string,
  claim_name: string,
  cred_def_id: string,
  credential_attrs: {
    [string]: string,
  },
  from_did: string,
  libindy_offer: string,
  msg_ref_id: ?string,
  msg_type: string,
  schema_seq_no: number,
  to_did: string,
  version: string,
|}

export type SerializedClaimOfferData = {
  agent_did: null,
  agent_vk: null,
  cred_id: null,
  credential: null,
  credential_name: null,
  credential_offer: ?MessageClaimOfferDetails,
  credential_request: null,
  msg_uid: null,
  my_did: null,
  my_vk: null,
  payment_info: ?MessagePaymentDetails,
  payment_txn: null,
  source_id: string,
  state: number,
  their_did: null,
  their_vk: null,
}

export type MessagePaymentDetails = {|
  payment_addr: string,
  payment_required: string,
  price: number,
|}

export type ParsedDecryptedPayloadMessage = Array<
  MessageClaimOfferDetails | MessagePaymentDetails
>
export type ParsedDecryptedPayload = {
  '@msg': string,
  '@type': {
    fmt: string,
    name: string,
    ver: string,
  },
}

export const ACKNOWLEDGE_MESSAGES_FAIL = 'ACKNOWLEDGE_MESSAGES_FAIL'
export const GET_MESSAGES_FAIL = 'GET_MESSAGES_FAIL'
export const ACKNOWLEDGE_MESSAGES = 'ACKNOWLEDGE_MESSAGES'
export const GET_MESSAGES_SUCCESS = 'GET_MESSAGES_SUCCESS'
export const GET_MESSAGES_LOADING = 'GET_MESSAGES_LOADING'
export const GET_UN_ACKNOWLEDGED_MESSAGES = 'GET_UN_ACKNOWLEDGED_MESSAGES'

export type GetUnacknowledgedMessagesAction = {
  type: typeof GET_UN_ACKNOWLEDGED_MESSAGES,
}
export type GetMessagesLoadingAction = {
  type: typeof GET_MESSAGES_LOADING,
}

export type GetMessagesSuccessAction = {
  type: typeof GET_MESSAGES_SUCCESS,
}

export type AcknowledgeMessagesAction = {
  type: typeof ACKNOWLEDGE_MESSAGES,
}

export type GetMessagesFailAction = {
  type: typeof GET_MESSAGES_FAIL,
}

export type AcknowledgeMessagesFailAction = {
  type: typeof ACKNOWLEDGE_MESSAGES_FAIL,
  error: string,
}
