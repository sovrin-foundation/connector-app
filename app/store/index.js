// @flow
import { createStore, combineReducers, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import { all } from 'redux-saga/effects'
import logger from 'redux-logger'
import user, { watchUserStore } from './user/user-store'
import pushNotification, {
  watchPushNotification,
} from '../push-notification/push-notification-store'
import connections, { watchConnection } from './connections-store'
import config, { watchConfig, watchGetMessagesSaga } from './config-store'
import authentication, {
  watchAuthentication,
} from '../authentication/authentication-store'
import deepLink from '../deep-link/deep-link-store'
import route from './route-store'
import lock, {
  watchLock,
  watchPressEventInLockSelectionScreen,
  watchEnableTouchId,
  watchDisableTouchId,
} from '../lock/lock-store'
import smsPendingInvitation, {
  watchSmsPendingInvitationSaga,
} from '../sms-pending-invitation/sms-pending-invitation-store'
import claimOffer, { watchClaimOffer } from '../claim-offer/claim-offer-store'
import proofRequest, {
  watchProofRequestAccepted,
  watchPersistProofRequests,
  watchProofRequestReceived,
} from '../proof-request/proof-request-store'
import invitation, { watchInvitation } from '../invitation/invitation-store'
import claim, { watchClaim } from '../claim/claim-store'
import proof, { watchProof } from '../proof/proof-store'
import history, {
  watchConnectionHistory,
} from '../connection-history/connection-history-store'
import historyRecorder from '../connection-history/history-middleware'
import wallet, { watchWalletStore } from '../wallet/wallet-store'
import eula, { watchEula } from '../eula/eula-store'
import restore, { watchRestore } from '../restore/restore-store'
import backup, { watchBackup } from '../backup/backup-store'
import onfido, { watchOnfido } from '../onfido/onfido-store'
import offline, { watchOffline } from '../offline/offline-store'
import { hydrate } from './hydration-store'
import {
  watchLedgerStore,
  ledgerStoreReducer as ledger,
} from './ledger/ledger-store'

const sagaMiddleware = createSagaMiddleware()

const appReducer = combineReducers({
  config,
  connections,
  deepLink,
  authentication,
  pushNotification,
  route,
  smsPendingInvitation,
  user,
  lock,
  claimOffer,
  proofRequest,
  invitation,
  claim,
  proof,
  history,
  wallet,
  eula,
  restore,
  backup,
  ledger,
  offline,
  onfido,
})

let middlewares = [historyRecorder]

if (process.env.NODE_ENV !== 'test' && __DEV__) {
  // skip logger middleware if we are running tests
  middlewares.push(logger)
}

middlewares.push(sagaMiddleware)

const store = createStore(appReducer, applyMiddleware(...middlewares))

// $FlowFixMe Don't know how to fix polymorphic type coercion
sagaMiddleware.run(function*() {
  return yield all([
    watchConnection(),
    watchConfig(),
    watchAuthentication(),
    watchLock(),
    watchSmsPendingInvitationSaga(),
    watchClaimOffer(),
    watchPushNotification(),
    watchInvitation(),
    watchClaim(),
    watchPressEventInLockSelectionScreen(),
    watchEnableTouchId(),
    watchDisableTouchId(),
    watchProof(),
    watchProofRequestAccepted(),
    watchConnectionHistory(),
    watchUserStore(),
    watchWalletStore(),
    watchBackup(),
    watchEula(),
    watchRestore(),
    hydrate(),
    watchGetMessagesSaga(),
    watchPersistProofRequests(),
    watchProofRequestReceived(),
    watchLedgerStore(),
    watchOffline(),
  ])
})

export default store
