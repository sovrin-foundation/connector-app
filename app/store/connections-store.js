// @flow
import {
  put,
  takeLatest,
  takeEvery,
  call,
  select,
  all,
} from 'redux-saga/effects'
import { secureSet, secureDelete, getHydrationItem } from '../services/storage'
import { CONNECTIONS } from '../common'
import {
  getAgencyUrl,
  getAgencyVerificationKey,
  getUserOneTimeInfo,
  getPoolConfig,
  getAllConnection,
  getThemes,
  getConnection as getConnectionBySenderDid,
} from './store-selector'
import { color } from '../common/styles/constant'
import { bubbleSize } from '../common/styles'
import type { CustomError, GenericObject } from '../common/type-common'
import type {
  ConnectionStore,
  Connection,
  Connections,
  ConnectionThemes,
} from './type-connection-store'
import {
  NEW_CONNECTION,
  DELETE_CONNECTION_SUCCESS,
  DELETE_CONNECTION_FAILURE,
  DELETE_CONNECTION,
  STORAGE_KEY_THEMES,
  HYDRATE_CONNECTION_THEMES,
} from './type-connection-store'
import type {
  DeleteConnectionSuccessEventAction,
  DeleteConnectionFailureEventAction,
  DeleteConnectionEventAction,
} from './type-connection-store'
import {
  deleteConnection,
  getHandleBySerializedConnection,
} from '../bridge/react-native-cxs/RNCxs'
import { RESET } from '../common/type-common'
import type { UserOneTimeInfo } from './user/type-user-store'
import { promptBackupBanner } from '../backup/backup-store'
import { HYDRATED } from './type-config-store'
import { captureError } from '../services/error/error-handler'

const UPDATE_CONNECTION_THEME = 'UPDATE_CONNECTION_THEME'
export const NEW_CONNECTION_SUCCESS = 'NEW_CONNECTION_SUCCESS'
export const UPDATE_STATUS_BAR_THEME = 'UPDATE_STATUS_BAR_THEME'
const NEW_CONNECTION_FAIL = 'NEW_CONNECTION_FAIL'
const HYDRATE_CONNECTIONS = 'HYDRATE_CONNECTIONS'

const initialState: ConnectionStore = {
  data: {},
  isFetching: false,
  isPristine: true,
  connectionThemes: {
    default: {
      primary: `rgba(${color.actions.button.primary.rgba})`,
      secondary: `rgba(${color.actions.button.secondary.rgba})`,
    },
  },
  claimMap: null,
  error: {
    code: '',
    message: '',
  },
  hydrated: false,
}

// TODO:KS As of now we have added flow to this file and only checking imports
// but we need to fix all any types. I will do that once claims are done

export const connectionMapper = ({
  logoUrl,
  size = bubbleSize.XL,
  senderName = 'Unknown',
  ...otherArgs
}: GenericObject) => ({
  logoUrl,
  size,
  senderName,
  ...otherArgs,
})

export const updateConnectionTheme = (
  logoUrl: string,
  primaryColor: string,
  secondaryColor: string
) => ({
  type: UPDATE_CONNECTION_THEME,
  logoUrl,
  primaryColor,
  secondaryColor,
})

export const saveNewConnection = (connection: GenericObject) => ({
  type: NEW_CONNECTION,
  connection,
})

export const saveNewConnectionSuccess = (connection: GenericObject) => ({
  type: NEW_CONNECTION_SUCCESS,
  connection,
})

//TODO refactor create a new store for ui
export const updateStatusBarTheme = (statusColor?: string) => ({
  type: UPDATE_STATUS_BAR_THEME,
  color: statusColor || color.bg.tertiary.color,
})

export const saveNewConnectionFailed = (error: CustomError) => ({
  type: NEW_CONNECTION_FAIL,
  error,
})

export const deleteConnectionAction = (
  senderDID: string
): DeleteConnectionEventAction => ({
  type: DELETE_CONNECTION,
  senderDID,
})

export function* deleteConnectionOccurredSaga(
  action: DeleteConnectionEventAction
): Generator<*, *, *> {
  const connections: GenericObject = yield select(getAllConnection)

  const [connection]: Array<Connection> = yield select(
    getConnectionBySenderDid,
    action.senderDID
  )
  const { [connection.myPairwiseDid]: deleted, ...rest } = connections

  const connectionHandle = yield call(
    getHandleBySerializedConnection,
    connection.vcxSerializedConnection
  )

  try {
    yield call(secureSet, CONNECTIONS, JSON.stringify(rest))
    yield put(deleteConnectionSuccess(rest))
    yield call(deleteConnection, connectionHandle)
  } catch (e) {
    captureError(e)
    yield put(deleteConnectionFailure(connection, e))
  }
}

export function* watchDeleteConnectionOccurred(): any {
  yield takeLatest(DELETE_CONNECTION, deleteConnectionOccurredSaga)
}

export function* loadNewConnectionSaga(
  action: GenericObject
): Generator<*, *, *> {
  const {
    identifier,
    logoUrl,
    senderDID,
    senderEndpoint,
    senderName,
    myPairwiseDid,
    myPairwiseVerKey,
    myPairwiseAgentDid,
    myPairwiseAgentVerKey,
    myPairwisePeerVerKey,
    vcxSerializedConnection,
  }: Connection = action.connection.newConnection

  try {
    const connection = {
      identifier,
      logoUrl,
      senderDID,
      senderEndpoint,
      senderName,
      myPairwiseDid,
      myPairwiseVerKey,
      myPairwiseAgentDid,
      myPairwiseAgentVerKey,
      myPairwisePeerVerKey,
      vcxSerializedConnection,
    }

    yield put(promptBackupBanner(true))
    yield put(saveNewConnectionSuccess(connection))
    yield* persistConnections()
  } catch (e) {
    yield put(saveNewConnectionFailed(e))
  }
}

export function* watchNewConnection(): any {
  yield takeEvery(NEW_CONNECTION, loadNewConnectionSaga)
}

export function* persistConnections(): Generator<*, *, *> {
  try {
    const connections = yield select(getAllConnection)
    yield call(secureSet, CONNECTIONS, JSON.stringify(connections))
  } catch (e) {
    captureError(e)
    console.log(`hydrateConnectionSaga: ${e}`)
  }
}

export const hydrateConnections = (connections: Connections) => ({
  type: HYDRATE_CONNECTIONS,
  connections,
})

export function* hydrateConnectionSaga(): Generator<*, *, *> {
  try {
    const connections = yield call(getHydrationItem, CONNECTIONS)
    if (connections) {
      yield put(hydrateConnections(JSON.parse(connections)))
    }
  } catch (e) {
    // to capture secure get
    captureError(e)
    console.log(`hydrateConnectionSaga: ${e}`)
  }
}

export const getConnections = (connectionsData: ?Connections) =>
  connectionsData ? Object.values(connectionsData) : []

export const getConnection = (
  remoteConnectionId: string,
  connections: Connections
) =>
  Object.values(connections).filter(function(c: any) {
    return c.remoteConnectionId === remoteConnectionId
  })

export const getConnectionLogo = (logoUrl: ?string) =>
  logoUrl ? { uri: logoUrl } : require('../images/logo_sovrin.png')

export const deleteConnectionSuccess = (
  filteredConnections: Connections
): DeleteConnectionSuccessEventAction => ({
  type: DELETE_CONNECTION_SUCCESS,
  filteredConnections,
})

export const deleteConnectionFailure = (
  connection: Connection,
  error: CustomError
): DeleteConnectionFailureEventAction => ({
  type: DELETE_CONNECTION_FAILURE,
  connection,
  error,
})

export const hydrateConnectionThemes = (themes: ConnectionThemes) => ({
  type: HYDRATE_CONNECTION_THEMES,
  themes,
})

export function* persistThemes(): Generator<*, *, *> {
  const themes = yield select(getThemes)
  try {
    yield call(secureSet, STORAGE_KEY_THEMES, JSON.stringify(themes))
  } catch (e) {
    // capture error for secure set
    captureError(e)
    console.error(`persistThemes: ${e}`)
  }
}

export function* hydrateThemes(): Generator<*, *, *> {
  try {
    const themes = yield call(getHydrationItem, STORAGE_KEY_THEMES)
    if (themes) {
      yield put(hydrateConnectionThemes(JSON.parse(themes)))
    }
  } catch (e) {
    // capture error for secure get
    captureError(e)
    console.error(`hydrateThemes: ${e}`)
  }
}

export function* removePersistedThemes(): Generator<*, *, *> {
  try {
    yield call(secureDelete, STORAGE_KEY_THEMES)
  } catch (e) {
    // capture error for secure delete
    captureError(e)
    console.error(`removePersistedThemes: ${e}`)
  }
}

export function* watchUpdateConnectionTheme(): any {
  yield takeLatest(UPDATE_CONNECTION_THEME, persistThemes)
}

export function* watchConnection(): any {
  yield all([
    watchDeleteConnectionOccurred(),
    watchNewConnection(),
    watchUpdateConnectionTheme(),
  ])
}

export default function connections(
  state: ConnectionStore = initialState,
  action: any
) {
  switch (action.type) {
    case UPDATE_CONNECTION_THEME:
      return {
        ...state,
        connectionThemes: {
          ...state.connectionThemes,
          [action.logoUrl]: {
            primary: action.primaryColor,
            secondary: action.secondaryColor,
          },
        },
      }
    case HYDRATE_CONNECTION_THEMES:
      return {
        ...state,
        connectionThemes: {
          ...state.connectionThemes,
          ...action.themes,
        },
      }
    case NEW_CONNECTION:
      return {
        ...state,
        isFetching: true,
        isPristine: false,
        error: initialState.error,
      }
    case NEW_CONNECTION_SUCCESS:
      const { connection, connection: { identifier } } = action
      return {
        ...state,
        isFetching: false,
        data: {
          ...state.data,
          [identifier]: connectionMapper(connection),
        },
      }
    case NEW_CONNECTION_FAIL:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    case DELETE_CONNECTION_SUCCESS:
      const filteredData = { ...action.filteredConnections }
      return {
        ...state,
        data: filteredData,
      }
    case UPDATE_STATUS_BAR_THEME:
      return {
        ...state,
        statusBarTheme: action.color,
      }
    case HYDRATE_CONNECTIONS:
      return {
        ...state,
        data: action.connections,
      }
    case HYDRATED:
      return {
        ...state,
        hydrated: true,
      }
    case RESET:
      return initialState
    default:
      return state
  }
}
