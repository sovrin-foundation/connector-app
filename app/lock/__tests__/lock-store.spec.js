// @flow
import { put, call } from 'redux-saga/effects'
import type { LockStore } from '../type-lock'
import {
  SET_PIN,
  CHECK_PIN,
  CHECK_PIN_IDLE,
  LOCK_ENABLE,
  PIN_ENABLED_KEY,
  IN_RECOVERY,
  PIN_HASH,
  SALT,
} from '../type-lock'
import lockReducer, {
  addPendingRedirection,
  clearPendingRedirect,
  setPin,
  lockEnable,
  checkPin,
  checkPinSuccess,
  checkPinFail,
  setPinAction,
  checkPinStatusIdle,
  checkPinAction,
  setInRecovery,
} from '../lock-store'
import {
  secureSet,
  secureGet,
  safeSet,
  safeGet,
  getHydrationItem,
} from '../../services/storage'
import { generateSalt, pinHash } from '../pin-hash'
import { VCX_INIT_SUCCESS } from '../../store/type-config-store'
import { expectSaga } from 'redux-saga-test-plan'
import * as matchers from 'redux-saga-test-plan/matchers'
import { throwError } from 'redux-saga-test-plan/providers'
import { configStoreHydratedInstalledVcxInitSuccess } from '../../../__mocks__/data/config-store-mock-data'

const initialState: LockStore = {
  pendingRedirection: null,
  checkPinStatus: CHECK_PIN_IDLE,
  // we are assuming that app will be locked by default
  // and it will be unlocked either when user set security first time
  // or user unlock the app every time user opens the app
  // this property needs to be set accordingly
  isAppLocked: true,
  showDevMode: false,
  isLockEnabled: 'false',
  isTouchIdEnabled: false,
  inRecovery: 'false',
}

describe('LockStore', () => {
  it('add pending redirection change state correctly', () => {
    const expectedState = lockReducer(
      initialState,
      addPendingRedirection([{ routeName: 'home' }])
    )
    expect(expectedState).toMatchSnapshot()
  })

  it('clear pending redirect if clear action is raised', () => {
    const pendingState = lockReducer(
      initialState,
      addPendingRedirection([{ routeName: 'home' }])
    )
    const expectedState = lockReducer(pendingState, clearPendingRedirect())
    expect(expectedState).toMatchSnapshot()
  })

  it('set pin should work fine', () => {
    const pin = '123456'
    const salt = 'salt'
    const hashedPin = 'some-hash'

    return expectSaga(setPin, setPinAction(pin))
      .provide([
        [matchers.call.fn(generateSalt), salt],
        [matchers.call.fn(pinHash, pin, salt), hashedPin],
      ])
      .call(secureSet, PIN_HASH, hashedPin)
      .call(secureSet, SALT, salt)
      .call(safeSet, PIN_ENABLED_KEY, 'true')
      .call(safeSet, IN_RECOVERY, 'false')
      .put(lockEnable('true'))
      .run()
  })

  it('check pin flow should work if correct pin is passed', () => {
    const pin = '123456'
    const salt = 'salt'
    const expectedPinHash = 'expectedPinHash'

    return expectSaga(checkPin, checkPinAction(pin))
      .provide([
        [matchers.call.fn(safeGet, IN_RECOVERY), null],
        [call(getHydrationItem, SALT), salt],
        [call(getHydrationItem, PIN_HASH), expectedPinHash],
        [matchers.call.fn(pinHash, pin, salt), expectedPinHash],
      ])
      .put(checkPinSuccess())
      .run()
  })

  it('check pin flow should work if correct pin is passed in recovery', () => {
    const pin = '123456'
    const salt = 'salt'
    const expectedPinHash = 'expectedPinHash'

    return expectSaga(checkPin, checkPinAction(pin))
      .withState({ config: configStoreHydratedInstalledVcxInitSuccess })
      .provide([
        [matchers.call.fn(safeGet, IN_RECOVERY), 'true'],
        [call(getHydrationItem, SALT), salt],
        [call(getHydrationItem, PIN_HASH), expectedPinHash],
        [matchers.call.fn(pinHash, pin, salt), expectedPinHash],
      ])
      .call(safeSet, IN_RECOVERY, 'false')
      .put(setInRecovery('false'))
      .put(checkPinSuccess())
      .run()
  })

  it('check pin flow should fail if incorrect pin is passed', () => {
    const pin = '123456'
    const salt = 'salt'
    const expectedPinHash = 'expectedPinHash'
    const enteredPinHash = 'enteredPinHash'

    return expectSaga(checkPin, checkPinAction(pin))
      .provide([
        [matchers.call.fn(safeGet, IN_RECOVERY), null],
        [call(getHydrationItem, SALT), salt],
        [call(getHydrationItem, PIN_HASH), expectedPinHash],
        [matchers.call.fn(pinHash, pin, salt), enteredPinHash],
      ])
      .put(checkPinFail())
      .run()
  })

  it('should set checkPinStatus to idle', () => {
    const failedState = lockReducer(initialState, checkPinFail())
    expect(failedState).toMatchSnapshot()
    const expectedState = lockReducer(failedState, checkPinStatusIdle())
    expect(expectedState).toMatchSnapshot()
  })
})
