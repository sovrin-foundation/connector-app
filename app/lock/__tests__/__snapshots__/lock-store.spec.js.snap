// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LockStore add pending redirection change state correctly 1`] = `
Object {
  "checkPinStatus": "CHECK_PIN_IDLE",
  "inRecovery": "false",
  "isAppLocked": true,
  "isLockEnabled": "false",
  "isTouchIdEnabled": false,
  "pendingRedirection": Array [
    Object {
      "routeName": "home",
    },
  ],
  "showDevMode": false,
}
`;

exports[`LockStore clear pending redirect if clear action is raised 1`] = `
Object {
  "checkPinStatus": "CHECK_PIN_IDLE",
  "inRecovery": "false",
  "isAppLocked": true,
  "isLockEnabled": "false",
  "isTouchIdEnabled": false,
  "pendingRedirection": null,
  "showDevMode": false,
}
`;

exports[`LockStore should set checkPinStatus to idle 1`] = `
Object {
  "checkPinStatus": "CHECK_PIN_FAIL",
  "inRecovery": "false",
  "isAppLocked": true,
  "isLockEnabled": "false",
  "isTouchIdEnabled": false,
  "pendingRedirection": null,
  "showDevMode": false,
}
`;

exports[`LockStore should set checkPinStatus to idle 2`] = `
Object {
  "checkPinStatus": "CHECK_PIN_IDLE",
  "inRecovery": "false",
  "isAppLocked": true,
  "isLockEnabled": "false",
  "isTouchIdEnabled": false,
  "pendingRedirection": null,
  "showDevMode": false,
}
`;
